### PR TITLE
Componentwise mat mods

### PR DESCRIPTION
**Pull request by keckler**
## Description
This PR addresses #513  by adding the ability to specify material modifications on a component level. Using the `componentwise` attribute in the `material modifictions` section of the blueprints, a given material modification may be specified to apply to only a selection of components in the block.

Backwards compatibility is maintained in terms of the blueprints input file. But an additional attribute `blockwise` has been added for the user to highlight that a given material modification applies to all applicable components in the block. This `blockwise` keyword may be used optionally and material modifications without either the `componentwise` or `blockwise` keyword will be treated blockwise as before.

The user docs have been updated to reflect this change.

---

## Checklist

- [x] Tests have been added/updated to verify that the new or changed code works.
- [x] Docstrings were included and updated as necessary
- [x] The code is understandable and maintainable to people beyond the author
- [x] There is no commented out code in this PR.

If user exposed functionality was added/changed:

- [x] Documentation added/updated in the `doc` folder.
- [ ] New or updated dependencies have been added to `setup.py`.

### CODE REVIEW

**Comment by keckler** [comment]
Any comments on the changes to the input parsing would be particularly welcomed. Also the attribute names `blockwise` and `componentwise` could potentially be improved upon.

**Comment by keckler** [comment]
Reviewer needed. @john-science @ntouran @jakehader

**Comment by john-science** [comment]
@keckler Thanks!  I get back from vacation today and I'll take a look!

@ntouran I will want your opinion on this, as you have a better view of other material/geometry work coming up in the future.

**Comment by keckler** [comment]
Of course, no rush over the holidays!

File: armi/reactor/blueprints/assemblyBlueprint.py [hunk 1]
```diff
@@ -153,17 +192,67 @@ def _constructAssembly(self, cs, blueprint):
...
+        assemblyMaterialInput = {}
+
+        if self.materialModifications != None:
+            assemblyMaterialInput[""] = {}
+            for modName, modList in self.materialModifications.items():
+                assemblyMaterialInput[""][modName] = modList
+
+            if self.materialModifications.blockwise:
+                for modName, modList in self.materialModifications.blockwise.items():
+                    if modName in assemblyMaterialInput[""].keys():
```
Comment (ntouran) [inline armi/reactor/blueprints/assemblyBlueprint.py:226]: no need to say `.keys()` here. The `in` statement checks keys already.

File: armi/reactor/blueprints/assemblyBlueprint.py [hunk 2]
```diff
@@ -153,17 +192,67 @@ def _constructAssembly(self, cs, blueprint):
...
+        single value to apply to all blocks without needing to write out a whole list.
+
+        To maintain backwards compatibility, we first assumed that there might be
+        blockwise matMods on the base object, then we add in any blockwise matMods
+        that are explicitly denoted using the `blockwise` attribute.
+        """
+        assemblyMaterialInput = {}
+
+        if self.materialModifications != None:
+            assemblyMaterialInput[""] = {}
```
Comment (ntouran) [inline armi/reactor/blueprints/assemblyBlueprint.py:220]: I recommend defining a special module-level constant like `BLOCKWISE = "_BLOCKWISE"` and using that instead of `""` as the blockwise key. That will add some additional useful context.

File: armi/reactor/blueprints/assemblyBlueprint.py [hunk 3]
```diff
@@ -153,17 +192,67 @@ def _constructAssembly(self, cs, blueprint):
...
+                        )
+                    assemblyMaterialInput[""][modName] = modList
+
+            if self.materialModifications.componentwise:
+                for component, mods in self.materialModifications.componentwise.items():
+                    assemblyMaterialInput[component] = {}
+                    for modName, modList in mods.items():
+                        assemblyMaterialInput[component][modName] = modList
+
+        return assemblyMaterialInput
```
Comment (ntouran) [inline armi/reactor/blueprints/assemblyBlueprint.py:239]: Why not just modify the object's `self.materialModifications` data structure directly here rather than making a modified copy of the data and passing it around in function calls?

File: armi/reactor/blueprints/assemblyBlueprint.py [hunk 4]
```diff
@@ -174,15 +263,16 @@ def _createBlock(self, cs, blueprint, bDesign, axialIndex):
 
     def _checkParamConsistency(self, a):
         """Check that the number of block params specified is equal to the number of blocks specified."""
-        materialInputs = self.materialModifications or {}
+        materialInputs = self._compileAssemblyMaterialInput()
```
Comment (ntouran) [inline armi/reactor/blueprints/assemblyBlueprint.py:266]: It's surprising that a checkParamConsistency call would do data compilation step. I suggest reverting this line and calling `self._compileAssemblyMaterialInput()` at the higher level, right before `_checkParamConsistency` is called.

File: armi/reactor/blueprints/blockBlueprint.py [hunk 5]
```diff
@@ -170,6 +179,46 @@ def construct(
...
+        Get the blockwise material modifications and those specifically for this
+        component.
+
+        If a material modification is specified both blockwise and componentwise
+        for a given component, the componentwise value will be used.
+        """
+        filteredMaterialInput = {}
+
+        # first add the blockwise modifications without question
+        if "" in materialInput.keys():
```
Comment (ntouran) [inline armi/reactor/blueprints/blockBlueprint.py:204]: can omit `.keys()`

File: doc/user/inputs/blueprints.rst [hunk 6]
```diff
@@ -404,9 +404,98 @@ material modifications
...
+                xs types: [A]
+                material modifications:
+                    componentwise:
+                        fuel1:
+                            U235_wt_frac: [0.20]
+                        fuel2:
+                            Zr_wt_frac: [0.02]
+                    U235_wt_frac: [0.30]
+
+  This example would apply an enrichment of 20% to the `fuel1` component and an
```
Comment (ntouran) [inline doc/user/inputs/blueprints.rst:446]: `fuel1` and other literals should have double-backticks around it for proper rst rendering

File: doc/user/inputs/blueprints.rst [hunk 7]
```diff
@@ -404,9 +404,98 @@ material modifications
...
+  enrichment of 30% to all other components in the block that accept the `U235_wt_frac`
+  material modification. The user may alternatively use the `blockwise` attribute
+  to more clearly indicate that the material modification type is being applied to
+  all applicable components in the block::
+
+                material modifications:
+                    componentwise:
```
Comment (ntouran) [inline doc/user/inputs/blueprints.rst:453]: does `bycomponent`  and `byblock` sound better to you?  Or just `component` and `block`? The `wise` part in these names feels slightly funny to me. Not a big deal though.

**Review by ntouran — CHANGES_REQUESTED** [review]
Looks like a pretty reasonable solution to a legit inadequacy of the input. I have a few minor comments in here, and would like to get @john-science's take on it as well since it's a input change. 

I am a little concerned about how this will work when we have better treatment of blocks containing composites (collections of collections + components), but the overall concept will stand as long as we iterate over the final components in such scenarios. 

Thanks for the unit test and doc updates too!

File: armi/reactor/blueprints/assemblyBlueprint.py (hunk 2 above)
Comment (keckler) [reply armi/reactor/blueprints/assemblyBlueprint.py:220]: I struggled with coming up with a good key name for the blockwise material modifications, because I didn't want to use a name that might possibly conflict with a name that the user might have for a component. 

Like, if the user were to name a component `_BLOCKWISE` for whatever odd reason, then any material modifications to that component would be overwritten internally. I know it is unlikely that a user might use that exact name, but still... Trying to avoid conflicts here.

I figured the safest way would be to not use a real word at all, hence the use of `""`. I figured that yamlize wouldn't allow an empty string to be used as an attribute name, but now that I actually checked, I find out I was wrong. So I'll change it to something more clear.

File: armi/reactor/blueprints/tests/test_materialModifications.py [hunk 8]
```diff
@@ -76,6 +84,93 @@ def test_u235_wt_frac_modification(self):
...
+        assert_allclose(0.20, u235 / u)
+
+    def test_u235_wt_frac_componentwise_modification1(self):
+        a = self.loadUZrAssembly(
+            """
+        material modifications:
+            componentwise:
+                fuel1:
+                    U235_wt_frac: [0.20]
+            U235_wt_frac: [0.30]
```
Comment (jakehader) [inline armi/reactor/blueprints/tests/test_materialModifications.py:117]: Can you describe how this would work if both a general wt frac and a component wt frac are provided together? Does one invalidate the other or is there a check that we should have to fail or warn the user?

Also, maybe if we are going this route then does it even make sense to maintain block material modifications? Maybe from a backwards compatibility perspective, but I think that what we really want are just mods on the component level anyways. 

@ntouran / @john-science?

File: doc/user/inputs/blueprints.rst (hunk 7 above)
Comment (keckler) [reply doc/user/inputs/blueprints.rst:453]: Yeah, I like `bycomponent` better. I'll probably change the key to `by component`, since yamlize is capable of that and it is more clear.

Still don't quite like `by block`, though. To me, it is not clear how the `by block` concept is different from how the material modifications are required to be entered as a list with one value per block. I'm thinking a name more like `block default` would work better. That strongly implies the concept that the material modification will fall back to the `block default` value if a `by component` value does not supersede it.

What do you think?

File: armi/reactor/blueprints/tests/test_materialModifications.py (hunk 8 above)
Comment (keckler) [reply armi/reactor/blueprints/tests/test_materialModifications.py:117]: Good question @jakehader . I added a section in the docs that describes how such a situation would work. Check [here](https://github.com/terrapower/armi/pull/517/files#diff-457ae5e548d65f136e8364a5191df93f944ca0b0734e8338f8dc0a6bbd2256fb).

The componentwise values supplant the blockwise values, if both are provided. Any value not provided componentwise will fall back to the blockwise value. And if neither are provided, then it will fall back to whatever the material class does (i.e. either use some default in the material class or fail).

File: doc/user/inputs/blueprints.rst [hunk 9]
```diff
@@ -404,9 +404,98 @@ material modifications
...
+                height: [0.5, 0.5]
+                axial mesh points: [1, 1]
+                xs types: [A, A]
+                material modifications:
+                    componentwise:
+                        fuel1:
+                            U235_wt_frac: [0.20, ''] # <-- the U235_wt_frac for the second block will go to the blockwise value
+                        fuel2: # the U235_wt_frac for fuel2 component in both axial blocks will go to the blockwise value
+                            Zr_wt_frac: [0.02, ''] # <-- the Zr_wt_frac for the second block will default because there is no blockwise value
+                    blockwise:
```
Comment (jakehader) [inline doc/user/inputs/blueprints.rst:497]: This answers my question above. This is nice that it keeps backwards compatibility. 

My main question here is - Is the length of the component and block mat mod lists required to be the same length as the number of blocks in the assembly? I believe they should be, but if not then we should make sure so there aren't any mistakes on the mapping. 

Also, it would be nice if there was a material modification summary table that could be produced for the user in the `stdout`. This is probably a separate diff, but would definitely help me verify that material mods are being applied as expected.

File: armi/reactor/blueprints/tests/test_materialModifications.py (hunk 8 above)
Comment (keckler) [reply armi/reactor/blueprints/tests/test_materialModifications.py:117]: My opinion is that the blockwise values are still useful, just to provide that default behavior if a componentwise value isn't provided.

File: doc/user/inputs/blueprints.rst (hunk 9 above)
Comment (keckler) [reply doc/user/inputs/blueprints.rst:497]: Yeah, the lengths of the mat mods lists all have to be the same. If they are not, the code will crash (or once #515 is fixed, a good error will be printed).

Comment (keckler) [reply doc/user/inputs/blueprints.rst:497]: A summary table would definitely be useful. But I agree that it would be most appropriate in a separate PR.

Another thing that I think would be useful is a straight default mat mod value (not a list) on a given assembly so that the full list does not have to be written out in the case that all axial blocks take the same value. Obviously not needed, but could be convenient and reduce errors.

Comment (jakehader) [reply doc/user/inputs/blueprints.rst:497]: That's true. It really does beg the question of what the mat mods are really supposed to do though. 

I understand that they interact with the material classes and there are some allowable kwargs for modifications, but it seems that with a full assembly modification we might get into the territory of considering setting number densities of nuclides directly on the blocks after the assembly had been modified. For example, reducing coolant density.

Maybe for now we just stick to what it is doing with applying mods to the material classes and also consider the full assembly mods and other uses for mat mods with @ntouran separately.

File: doc/user/inputs/blueprints.rst (hunk 7 above)
Comment (jakehader) [reply doc/user/inputs/blueprints.rst:453]: I like the idea of just adding `by component` and then leaving the block material modifications at the top level. This will keep backwards compatibility with inputs.

Comment (keckler) [reply doc/user/inputs/blueprints.rst:453]: Yeah, that might be the best way to go. I always tend to like symmetry, so if there is a `by block` attribute, my gut wants there to be an alternative attribute as well. But just leaving the blockwise values on the `material modifications` level might be best.

Comment (keckler) [reply doc/user/inputs/blueprints.rst]: I changed the latest commit to use `by component` and got rid of the `blockwise` option entirely. Now all block default values are entered as they were before on the `material modifications` level.

File: armi/reactor/blueprints/assemblyBlueprint.py (hunk 1 above)
Comment (keckler) [reply armi/reactor/blueprints/assemblyBlueprint.py]: Fixed.

File: armi/reactor/blueprints/assemblyBlueprint.py (hunk 2 above)
Comment (keckler) [reply armi/reactor/blueprints/assemblyBlueprint.py]: I changed the key.

File: armi/reactor/blueprints/assemblyBlueprint.py (hunk 3 above)
Comment (keckler) [reply armi/reactor/blueprints/assemblyBlueprint.py]: I changed the flow to get rid of the `_compileAssemblyMaterialInput` method entirely. So no duplicates are made and nothing is passed around anymore.

File: armi/reactor/blueprints/assemblyBlueprint.py (hunk 4 above)
Comment (keckler) [reply armi/reactor/blueprints/assemblyBlueprint.py]: I got rid of the `_compileAssemblyMaterialInput` method and changed this part of the code to just pull directly from `self.materialModifications` as needed.

File: armi/reactor/blueprints/blockBlueprint.py (hunk 5 above)
Comment (keckler) [reply armi/reactor/blueprints/blockBlueprint.py]: Fixed.

File: doc/user/inputs/blueprints.rst (hunk 6 above)
Comment (keckler) [reply doc/user/inputs/blueprints.rst]: Fixed.

**Comment by keckler** [comment]
I think that all of @ntouran and @jakehader 's comments have been addressed by the latest commit. If you have any other comments on the latest version, I'd be happy to hear them. And then also waiting to hear if @john-science has anything he wants to add.

File: armi/reactor/blueprints/assemblyBlueprint.py (hunk 3 above)
Comment (john-science) [reply armi/reactor/blueprints/assemblyBlueprint.py]: Since the changes were made (possibly to address the comments directly), I can't really see or understand this conversation any more.  Keckler's comment makes it sound like a good change was made here.  One of you two should either continue this conversation or "resolve" it.

File: armi/reactor/blueprints/assemblyBlueprint.py [hunk 10]
```diff
@@ -174,15 +220,19 @@ def _createBlock(self, cs, blueprint, bDesign, axialIndex):
...
             "xs types": self.xsTypes,
         }
-        for modName, modList in materialInputs.items():
-            paramName = "material modifications for {}".format(modName)
-            paramsToCheck[paramName] = modList
+
+        for mod in [self.materialModifications] + [
+            self.materialModifications.byComponent[key]
+            for key in self.materialModifications.byComponent.keys()
+        ]:
```
Comment (john-science) [inline armi/reactor/blueprints/assemblyBlueprint.py:232]: Unless I'm missing something, I believe this might be more clearly written as:

```python
        for mod in [self.materialModifications] + list(self.materialModifications.byComponent.values()):
```

File: doc/user/inputs/blueprints.rst (hunk 7 above)
Comment (john-science) [reply doc/user/inputs/blueprints.rst]: It looks to me like this conversation can be "resolve"d.  @keckler ?

File: doc/user/inputs/blueprints.rst (hunk 9 above)
Comment (john-science) [reply doc/user/inputs/blueprints.rst]: > Yeah, the lengths of the mat mods lists all have to be the same. If they are not, the code will crash (or once #515 is fixed, a good error will be printed).

Do we have a unit test for that?  If not, can we get one?

File: armi/reactor/blueprints/blockBlueprint.py [hunk 11]
```diff
@@ -170,6 +179,46 @@ def construct(
...
+            if component != "byBlock":
+                if component not in [componentDesign.name for componentDesign in self]:
+                    if materialInput[component]:  # ensure it is not empty
+                        raise ValueError(
+                            f"The component '{component}' used to specify a by-component"
+                            f" material modification is not in block '{self.name}'."
+                        )
+
+    @staticmethod
+    def _filterMaterialInput(materialInput, componentDesign):
```
Comment (john-science) [inline armi/reactor/blueprints/blockBlueprint.py:193]: Two thoughts:

1. If this is a `@staticmethod` then we want people to use this method outside this class, so it shouldn't start with an `_`.
2. This method seems cool, and I like it, but it also does a lot of work. Can we have a unit test or three on this method alone?

File: armi/reactor/blueprints/blockBlueprint.py [hunk 12]
```diff
@@ -170,6 +179,46 @@ def construct(
                 runLog.warning(str(e), single=True)
         return b
 
+    def _checkByComponentMaterialInput(self, materialInput):
```
Comment (john-science) [inline armi/reactor/blueprints/blockBlueprint.py:182]: Can we have a unit test for this method?

File: armi/reactor/blueprints/assemblyBlueprint.py [hunk 13]
```diff
@@ -174,15 +220,19 @@ def _createBlock(self, cs, blueprint, bDesign, axialIndex):
 
     def _checkParamConsistency(self, a):
```
Comment (john-science) [inline armi/reactor/blueprints/assemblyBlueprint.py:221]: It is somewhat unfair to you, @keckler , this method didn't have unit tests to begin with. But now I'm wondering if we need one to validate your change works.

File: armi/reactor/blueprints/assemblyBlueprint.py (hunk 10 above)
Comment (keckler) [reply armi/reactor/blueprints/assemblyBlueprint.py:232]: Of course, will fix and put back up.

**Comment by john-science** [comment]
I would like some more unit tests for this PR.

Having said that, I think the feature is sorely needed.  And I love that documentation was added.

Part of the reason I like the idea of unit tests on these new methods here is I like the idea of this PR a lot, and can imagine there are edge cases here that might cause problems that I can't think of.

It's solid work.

File: doc/user/inputs/blueprints.rst (hunk 7 above)
Comment (keckler) [reply doc/user/inputs/blueprints.rst]: In my view it is resolved, just didn't mark it as such in case anyone else felt differently.

File: armi/reactor/blueprints/assemblyBlueprint.py (hunk 3 above)
Comment (keckler) [reply armi/reactor/blueprints/assemblyBlueprint.py]: Resolving now.

File: armi/reactor/blueprints/blockBlueprint.py (hunk 12 above)
Comment (keckler) [reply armi/reactor/blueprints/blockBlueprint.py:182]: Added.

File: armi/reactor/blueprints/blockBlueprint.py (hunk 11 above)
Comment (keckler) [reply armi/reactor/blueprints/blockBlueprint.py:193]: The reason for the `@staticmethod` is just to avoid passing the instance to the method, since the instance is not needed. But I don't expect this method to be used from outside the class, hence the underscore. Per our offline discussion, I am going to leave that unchanged.

Unit test added.

File: doc/user/inputs/blueprints.rst (hunk 9 above)
Comment (keckler) [reply doc/user/inputs/blueprints.rst]: Unit tests are added for this function. I also went ahead and fixed #515 myself in addressing this, since the two were so tied up with each other.

File: armi/reactor/blueprints/assemblyBlueprint.py (hunk 13 above)
Comment (keckler) [reply armi/reactor/blueprints/assemblyBlueprint.py]: Unit tests are added. Also went ahead and fixed #515 in the process, since it was so related.

**Comment by keckler** [comment]
Okay @john-science , I think all comments are addressed. Added a bunch of unit tests. Let me know if anything else looks amiss.

File: armi/reactor/blueprints/tests/test_assemblyBlueprints.py [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+    def loadCustomAssembly(self, assemblyInput):
+        yamlString = assemblyInput
+        design = blueprints.Blueprints.load(yamlString)
+        design._prepConstruction(settings.getMasterCs())
+        return design.assemblies["fuel a"]
+
+    def test_checkParamConsistency(self):
+        # make sure a good example doesn't error
+        a = self.loadCustomAssembly(self.twoBlockInput_correct)
```
Comment (ntouran) [inline armi/reactor/blueprints/tests/test_assemblyBlueprints.py:188]: this should technically have an assert statement afterwards to show not only that it runs but also that it loads values as expected

**Review by ntouran — APPROVED** [review]
Looks good to me! Could use one or two more asserts in the positive tests but NBD.

**Comment by ntouran** [comment]
@john-science  I'm good on this. You have any other requests? I'll let you squash/merge it if not.

File: armi/reactor/blueprints/tests/test_assemblyBlueprints.py (hunk 14 above)
Comment (john-science) [reply armi/reactor/blueprints/tests/test_assemblyBlueprints.py:188]: Ah, good catch!

I agree.  Pull the most important things out of that assembly and make sure they are correct.

**Comment by john-science** [comment]
> @john-science I'm good on this. You have any other requests? I'll let you squash/merge it if not.

I agree. Other than your comment about asserts in that test, this is ready. 

Nice work, @keckler !

**Comment by keckler** [comment]
@john-science I just added the assert statements into the test in the latest push.

**Comment by keckler** [comment]
Not rushing anybody, just wanted to make sure there is nothing more that needs to be done from me on this?
@john-science @ntouran
